standards/core/rln-contract.md

```diff
@@ -0,0 +1,209 @@
+- (TBD) freeze certain functionality (e.g., in case of an attack, stop new registrations, deposit withdrawals, etc).
+
+In the initial deployment:
+- there is no slashing;
+- membership expiration includes a grace period;
+- a user can extend their membership without putting up any additional deposit.
+
+### Membership life-cycle
+
+To register a membership, a user locks up a deposit.
+Each membership has an expiration date.
+After a membership expires, it enters a grace period, during which the user can extend it.
+Extending a membership requires no additional deposit.
+After the grace period, an expired membership can be taken over by another user.
+At any point, a user can withdraw their deposit and terminate the membership.
+
+In more detail, the membership life-cycle is as follows:
+1. A user registers a membership:
+    1. If there are expired unclaimed slots in the tree, use the slot with the earliest expiration date past its grace period, otherwise create a new slot;
+    2. if the membership re-uses a previously used slot, the previous user can claim their deposit back.
+2. The user sends messages while the membership is active:
+    1. If the user exceeds the rate limit, their over-limit messages are not propagated in that epoch; the deposit is not slashed.
+3. After the membership expires, the user can do one of the following:
+    1. Do nothing. The deposit remains in the contract, and the user retains the ability to send messages (respecting the per-epoch rate limit); until another user takes their slot
+    2. Withdraw the deposit. The user receives the full deposit back, and loses the ability to send messages;
+    3. Extend the membership. The user sends a transaction to re-enable their membership for another expiration term. The user only covers gas costs and does not have to provide additional funds. The original deposit remains in the contract. The membership is prolonged for another expiration term under the same conditions.
+
+The user can extend the membership both during and after its grace period.
+By not extending the membership before the grace period expires, the user assumes the risk of the membership being taken over at any moment.
+
+A user can hold multiple memberships.
+Memberships are not transferable.
+There is no limitation on the number of membership per user (apart from the global limits, see parameter table).
+
+## Contract governance and mutability
+
+The contract governance is as follows:
+
+1. the first deployment of the contract allows the `Owner` to modify certain parameters (TBD) under certain constraints (TBD);
+2. at some point, the `Onwer` would renounce their privileges, and the contract will become immutable;
+3. further upgrades, if necessary, can be done by deploying a new contract and migrating the membership set.
+
+## Parameters
+
+We make the following design choices:
+- there are `3` membership tiers: low-, mid-, and high-tier;
```

Comment:
Sounds reasonable to me. I'd consider two questions here:

1. What is simpler from the UI perspective: to offer one of three options to a user, or to let them pick a point inside an interval? Perhaps, we could do both: offer three default values and also support custom rate limits (under "advanced settings" or something) - similar to how wallets let you choose your gas price.
2. What would be simpler to implement - three tiers with hard-coded rate limits, or letting a user choose any point inside an interval, and calculate the membership price by multiplying a hard-coded unit price (the price of 1 message / epoch) by the chosen rate limit? ( @richard-ramos , any thoughts? )

Reply:
> What is simpler from the UI perspective: to offer one of three options to a user, or to let them pick a point inside an interval?

This question could be left to the UI implementation. Giving more flexibility to the developer.

I think it's a good idea to give flexibility here. Because with the flexibility, we can let the app dev decides what rate limit they want to offer to their user (or buy for their users).

The main question will be upgradability:
- With 3 tiers, we could change the prices on the same contract, and move from a linear to an exponential model. We can also decide if we hardcoded 3 tiers. Or use a map so we can have N tiers (and start with 3).
- with a `SOME_CONST` multiplier, we can change the const, but we won't be able to change the curve. OR, we actually have 2 parameters:
  - the constant
  -  the curve, ie, arithmetic operator we apply. Can be multiplier at first, then moved to log or exp etc

@richard-ramos , thoughts?

Reply:
I'm leaning towards a curve-based solution as more generalizable and updatable.

Let's define a function that takes as input the desired rate limit and outputs its price. Initially, that function will just multiply the rate limit by a constant. Later, we could replace it with a function consisting of multiple linear intervals (e.g., rates under `R1` cost `X1` per unit, rates between `R1` and `R2` cost `X2` per unit, etc, where `X1 > X2 > ...`). Ultimately, it could be a continious function like in AMMs, which for me looks like the most universal option.

Possible drawbacks:
- generalized price calculation will likely consume more gas than picking a price out of, say, three hard-coded prices for three tiers;
- I'm not sure how the _Owner_ would upgrade the price function. Can a variable be (re-)assigned a function value in Solidity? Ideally, an _Owner_-initiated price function upgrade would be something like `price_funcion = lambda x: sqrt(x)`; is this possible? If not, what's the alternative?

Reply:
> budget limit, i.e., 1ETH per month

It's not obvious to me how to add this feature to the current design. To clarify: as per the current spec spec, users don't pay per messages per se. A user locks up a deposit, the size of which depends on the rate limit they choose, and withdraws that deposit in full after their membership enters its grace period (i.e., "soft-expires"). In other words, the deposit doesn't get consumed by user's messages. In future versions of this spec we might consider only allowing partial deposit withdrawal (i.e., keeping part of the deposit as payment for the service), but that's out of scope for now.

> I consider it crucial to have a free tier

Wouldn't we open an attack vector here? We don't have an identity system, so with a free tier anyone would be able register a million memberships and overwhelm the network.

Reply:
> Wouldn't we open an attack vector here? We don't have an identity system, so with a free tier anyone would be able register a million memberships and overwhelm the network.

I don't think so. The registration process incurs charges to the user. I don't think anyone will register 1 million times.

Reply:
> The registration process incurs charges to the user.

Hm, doesn't the free tier imply free registration? It feels like we're talking about different things here, but I'm not sure where the misunderstanding comes from.

Reply:
From simplicity to implement, the easiest would be to have N tiers, but assuming the contract is upgradable, it should be possible to implement the curve mechanism and in case we want to change the curve, we'd upgrade the contract to use a different curve

Reply:
> in case we want to change the curve, we'd upgrade the contract

Does this imply that there would be a main contract that calls a different curve contract, and the latter would be replaced in case of an upgrade?

> I can try to implement an example of how would that look

Thanks, I don't think there is a need to implement a full example for now. I just want to make sure that the spec doesn't contain something that is overly complicated from the implementer's standpoint. Considering your comment, I'm leaning towards a curve-based mechanism with a linear curve function to start with. It looks more elegant to me than hard-coded tiers; and tiers can be implemented anyway as suggested rate limits on the application level.

Reply:
@Ivansete-status 

> The speed at which that user quota is spent depends on the usage. The more usage, the quickest the budget is consumed.

This is not how RLN work, no one track a quota per user in the system.

> Does this imply that there would be a main contract that calls a different curve contract, and the latter would be replaced in case of an upgrade?

Came here to suggest that. I am guessing the simplest way is:
- main contract has a field for "curve contract"address
- "curve contract"'s ABI is `f(a, b): c`
Then several curve contract can be deployed and admin changes the "curve contract address".

Would be interesting to check gas. a map might be the cheapest and simplest way (and most foolproof as well)